### PR TITLE
Build  LocationHierarchy Location hierarchy on sync strategy

### LINKED ISSUE #65: Add the ability to build the LocationHierarchy based on the Sync Strategy
## Context
- The LocationHierarchy endpoint currently expects the ID of the location we are trying to get the hierarchy for.  
- In some cases, this ID might need to be dynamic. This can be determined by the following:
    - **User assigned location**
        - Here, the LocationHierarchy endpoint will be triggered with the user-assigned locations as the location we get the LocationHierarchy for. 
    - **User selected location** 
        - Here the LocationHierarchy endpoint will be triggered with the user-selected locations as the locations to get the hierarchy for.  
        - This option can only be used if the following conditions are fulfilled:
        - The deployment/app user _Related Entity Location_ as its sync strategy
        - The `_syncLocations` parameter is not null from the request

## Acceptance Criteria
- [ ] Be able to build the hierarchy based on the login users location assignment
- [ ] Be able to build the hierarchy based on the user select locations
- [ ] Be able to buildthe hierarchy based on the provided location Id

## References
- https://docs.google.com/document/d/11Io0bMeQrEz4sdZYI4Qz-H4yRW8VYI-QJkNGt1qkiH4/edit?usp=sharing

### PR DESCRIPTION
**Pull request by lincmba**
fixes #65

**Engineer Checklist**

- [x] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [x] I have added documentation for any new feature(s) and configuration
      option(s) on the `README.md`
- [x] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [x] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [x] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 1]
```diff
@@ -4,22 +4,24 @@
...
+import java.util.stream.Collectors;
 
 import javax.annotation.Nullable;
 import javax.servlet.http.HttpServletRequest;
 
-import org.hl7.fhir.r4.model.Bundle;
-import org.hl7.fhir.r4.model.Location;
-import org.hl7.fhir.r4.model.Resource;
-import org.hl7.fhir.r4.model.StringType;
+import org.hl7.fhir.r4.model.*;
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:16]: Avoid wildcard imports

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java [hunk 2]
```diff
@@ -201,12 +205,12 @@ private boolean checkIfRoleExists(String roleName, List<String> existingRoles) {
         return existingRoles.contains(roleName);
     }
 
-    private IGenericClient createFhirClientForR4(FhirContext fhirContext) {
+    public static IGenericClient createFhirClientForR4(FhirContext fhirContext) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java:208]: I think we can move all these helper methods to the [Utils.java](https://github.com/onaio/fhir-gateway-extension/blob/main/plugins/src/main/java/org/smartregister/fhir/gateway/plugins/Utils.java) class

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java [hunk 3]
```diff
@@ -201,12 +205,12 @@ private boolean checkIfRoleExists(String roleName, List<String> existingRoles) {
...
         String fhirServer = System.getenv(Constants.PROXY_TO_ENV);
         return fhirContext.newRestfulGenericClient(fhirServer);
     }
 
-    private String getBinaryResourceReference(Composition composition) {
+    public static String getBinaryResourceReference(Composition composition) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java:213]: Same as above

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java [hunk 4]
```diff
@@ -240,7 +244,7 @@ private String getBinaryResourceReference(Composition composition) {
         return id;
     }
 
-    private Binary readApplicationConfigBinaryResource(
+    public static Binary readApplicationConfigBinaryResource(
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java:247]: SAB

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java [hunk 5]
```diff
@@ -264,7 +268,7 @@ private Composition readCompositionResource(String applicationId, FhirContext fh
         return compositionEntry != null ? (Composition) compositionEntry.getResource() : null;
     }
 
-    private String findSyncStrategy(Binary binary) {
+    public static String findSyncStrategy(Binary binary) {
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/PermissionAccessChecker.java:271]: SAB

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 6]
```diff
@@ -126,8 +136,109 @@ public List<Location> getDescendants(String locationId, Location parentLocation)
...
+        } else {
+            composition = null;
+        }
+        String binaryResourceReference =
+                PermissionAccessChecker.getBinaryResourceReference(composition);
+        Binary binary =
+                PermissionAccessChecker.readApplicationConfigBinaryResource(
+                        binaryResourceReference, fhirContext);
+        return PermissionAccessChecker.findSyncStrategy(binary);
+    }
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:231]: We can do some benchmarking here but I think there's an opportunity to optimise via caching by application ID. This will save us 2 API calls and the extra processing.

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 7]
```diff
@@ -161,7 +161,7 @@ public Bundle handleNonIdentifierRequest(
         List<String> selectedSyncLocations = extractSyncLocations(syncLocationsParam);
         String practitionerId = verifiedJwt.getSubject();
         List<String> userRoles = JwtUtils.getUserRolesFromJWT(verifiedJwt);
-        String syncStrategy = getSyncStrategy(verifiedJwt);
+        String syncStrategy = getSyncStrategyFromJwtToken(verifiedJwt);
```
Comment (ndegwamartin) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:164]: Nit pick (maybe) - I think we should have the method pass in the application id as parameter since conceptually we usually map an application id to a sync id.

Then call it something like like `getSyncStrategyByAppId(`

**Review by ndegwamartin — DISMISSED** [review]

**Review by dubdabasoduba — APPROVED** [review]
